### DISCUSSION TITLE
Process Manager Metadata for Commands

### DISCUSSION THREAD

**Discussion by arfl** [commanded/commanded#586] [Ideas]
I am curious: is it some way for process managers to add metadata to the dispatched commands. The `handle` function is supposed to return one command or a list of commands to be dispatched...

**Comment by slashdotdash** [comment]
Unfortunately it's not currently supported. 

Can you give me an example of what kind of metadata you want to include? We could extend to process manager API to allow setting metadata.

**Comment by trbngr** [comment]
One example: I'd love to be able to set the correlation id

**Comment by arfl** [comment]
We do not use the causation_id/correlation_id provided by commanded/eventstore because they are just uuids. Our IDs are URIs and look like this one: `tech.studitemps:my_bounded_context_name.my_event_name.e7f4e458-3da6-427d-88e3-227bbad22062`, where `studitemps.tech` is our web site. So just looking at this kind of IDs you can determine what kind of event it is and which bounded context has thrown this event.
So we would like to store those causation_ids/correlation_ids in process managers also.

**Comment by slashdotdash** [comment]
Pull request #389 provides access to enriched metadata in the [`Commanded.Event.Upcaster` protocol](https://hexdocs.pm/commanded/Commanded.Event.Upcaster.html). This could be used to copy any fields from the [`Commanded.EventStore.RecordedEvent`](https://hexdocs.pm/commanded/Commanded.EventStore.RecordedEvent.html) struct, such as the causation and correlation ids.

**Comment by jdav-dev** [comment]
My team has a use case for this.  We store a `user_id` in metadata.  We would like process managers to take this `user_id` from events it handles and add it to commands it dispatches.  Right now, process managers cannot add metadata, nor do they receive metadata with the events.  Our workaround is middleware that copies the `user_id` based on a command's `causation_id`.

**Comment by edwardzhou** [comment]
unable to access metadata not only in process manager, but also within Aggregate. Since Router.dispatch receives metadata in 3rd parameter, it is reasonable to pass it Aggregate.execute. this is very useful to carry the request user_id through the execution flow.

**Comment by nicklayb** [comment]
Similarly to @jdav-dev,  we need metadata to be forwarded to next commands. We are using a commanded middleware to authorize command originators to dispatch commands. Pretty basic middleware that checks if the command's metadata includes an originator and that the originator is allowed to perform such command. The thing is that the initial command's metadata aren't forwarded to subsequent therefore following commands aren't authorized.

I made some tests on [my fork here](https://github.com/nicklayb/commanded/commit/94715a846570c17835d9c37c86deadadb5a20bb0) to see what would make sense to fix it. I could make a PR and add tests if you feel it can be worth, otherwise, if you have recommendations on how it would be better, I would be happy to hear it.

**Comment by gtaranti** [comment]
I would also love this enhancement to the ProcessManager.
I can't think off any reason NOT to have the ability to add metadata at this point of the commands - event flow.

If not, we can add metadata as regular command/event data, but this bloats commands/events.

**Comment by brolund** [comment]
I would also like to have this feature!

**Comment by jvantuyl** [comment]
I have a much simpler need for this: "Controlling timeouts on dispatched commands individually."  Consistency would also be a nice one to set for some commands but not others.

Dispatching a command can accept some options (`metadata` being only one of them) and that list is only ever likely to grow.  At the heart of it, it's just dispatching using the normal function calls anyways ([here](https://github.com/commanded/commanded/blob/v1.4.2/lib/commanded/process_managers/process_manager_instance.ex#L410), I think).  Maybe just make the options generated [here](https://github.com/commanded/commanded/blob/v1.4.2/lib/commanded/process_managers/process_manager_instance.ex#L216) "default options" and allow overriding them?

Maybe we can just solve all of these problems forever with typing like:
```elixir
@callback handle(process_manager(), domain_event()) ::
  command()
  | {command(), dispatch_opts()}
  | [command() | {command(), dispatch_opts()}]
  | {:error, term()}
```
